File: config_server/ui/src/api/pipelineConfig.js
import { constructProtobufRequest, strToBytes, URL_PREFIX} from "@/api/common";
import userProto from "@/proto/user_pb";
import agentProto from "@/proto/agent_pb"


export async function listPipelineConfigs() {
    let url = URL_PREFIX + "ListPipelineConfigs"
    let req = new userProto.ListConfigsRequest()
    let res = await constructProtobufRequest(url, req, userProto.ListConfigsResponse)
    // res.configDetailsList.forEach(res => {
    //     res.detail =  base64ToStr(res.detail)
    // })
    return res
}

export async function getPipelineConfig(name){
    let url=URL_PREFIX+"GetPipelineConfig"
    let req = new userProto.GetConfigRequest()
    req.setConfigName(name)
    return await constructProtobufRequest(url, req, userProto.GetConfigResponse);
}

export async function createPipelineConfig(name,version,detail){
    let url=URL_PREFIX+"CreatePipelineConfig"
    let req = new userProto.CreateConfigRequest()
    let configDetail=new agentProto.ConfigDetail()
    configDetail.setName(name)
    configDetail.setVersion(version)
    configDetail.setDetail(strToBytes(detail))
    req.setConfigDetail(configDetail)
    return await constructProtobufRequest(url, req, userProto.CreateConfigResponse);
}

export async function updatePipelineConfig(name,version,detail){
    let url=URL_PREFIX+"UpdatePipelineConfig"
    let req = new userProto.UpdateConfigRequest()
    let configDetail=new agentProto.ConfigDetail()
    configDetail.setName(name)
    configDetail.setVersion(version)
    configDetail.setDetail(strToBytes(detail))
    req.setConfigDetail(configDetail)
    return await constructProtobufRequest(url, req, userProto.UpdateConfigResponse);
}

export async function deletePipelineConfig(name) {
    let url = URL_PREFIX + "DeletePipelineConfig"
    let req = new userProto.DeleteConfigRequest()
    req.setConfigName(name)
    return await constructProtobufRequest(url, req, userProto.DeleteConfigResponse);
}

export async function applyPipelineConfigToAgentGroup(configName,groupName){
    let url = URL_PREFIX + "ApplyPipelineConfigToAgentGroup"
    let req = new userProto.ApplyConfigToAgentGroupRequest()
    req.setConfigName(configName)
    req.setGroupName(groupName)
    return await constructProtobufRequest(url, req, userProto.ApplyConfigToAgentGroupResponse);
}

export async function removePipelineConfigFromAgentGroup(configName,groupName){
    let url = URL_PREFIX + "RemovePipelineConfigFromAgentGroup"
    let req = new userProto.RemoveConfigFromAgentGroupRequest()
    req.setConfigName(configName)
    req.setGroupName(groupName)
    return await constructProtobufRequest(url, req, userProto.RemoveConfigFromAgentGroupResponse);
}
export async function getAppliedAgentGroupsWithPipelineConfig(configName) {
    let url = URL_PREFIX + "GetAppliedAgentGroupsWithPipelineConfig"
    let req = new userProto.GetAppliedAgentGroupsRequest()
    req.setConfigName(configName)
    return await constructProtobufRequest(url, req, userProto.GetAppliedAgentGroupsResponse);
}

File: config_server/v2/service/test/request/pipeline_config.go
package request

import (
	"config-server/protov2"
	"fmt"
)

func CreatePipelineConfig(requestId []byte, detail *protov2.ConfigDetail) (*protov2.CreateConfigResponse, error) {
	url := fmt.Sprintf("%s/CreatePipelineConfig", UserUrlPrefix)
	req := &protov2.CreateConfigRequest{}
	res := &protov2.CreateConfigResponse{}
	req.RequestId = requestId
	req.ConfigDetail = detail

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func GetPipelineConfig(requestId []byte, name string) (*protov2.GetConfigResponse, error) {
	url := fmt.Sprintf("%s/GetPipelineConfig", UserUrlPrefix)
	req := &protov2.GetConfigRequest{}
	res := &protov2.GetConfigResponse{}
	req.RequestId = requestId
	req.ConfigName = name

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func GetAllPipelineConfig(requestId []byte) (*protov2.ListConfigsResponse, error) {
	url := fmt.Sprintf("%s/ListPipelineConfigs", UserUrlPrefix)
	req := &protov2.ListConfigsRequest{}
	res := &protov2.ListConfigsResponse{}
	req.RequestId = requestId

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func UpdatePipelineConfig(requestId []byte, detail *protov2.ConfigDetail) (*protov2.UpdateConfigResponse, error) {
	url := fmt.Sprintf("%s/UpdatePipelineConfig", UserUrlPrefix)
	req := &protov2.UpdateConfigRequest{}
	res := &protov2.UpdateConfigResponse{}
	req.RequestId = requestId
	req.ConfigDetail = detail

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func DeletePipelineConfig(requestId []byte, name string) (*protov2.DeleteConfigResponse, error) {
	url := fmt.Sprintf("%s/DeletePipelineConfig", UserUrlPrefix)
	req := &protov2.DeleteConfigRequest{}
	res := &protov2.DeleteConfigResponse{}
	req.RequestId = requestId
	req.ConfigName = name

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func ApplyPipelineConfig2AgentGroup(requestId []byte, configName string, agentGroupName string) (*protov2.ApplyConfigToAgentGroupResponse, error) {
	url := fmt.Sprintf("%s/ApplyPipelineConfigToAgentGroup", UserUrlPrefix)
	req := &protov2.ApplyConfigToAgentGroupRequest{}
	res := &protov2.ApplyConfigToAgentGroupResponse{}
	req.RequestId = requestId
	req.ConfigName = configName
	req.GroupName = agentGroupName

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func RemovePipelineConfigFromAgentGroup(requestId []byte, configName string, groupName string) (*protov2.RemoveConfigFromAgentGroupResponse, error) {
	url := fmt.Sprintf("%s/RemovePipelineConfigFromAgentGroup", UserUrlPrefix)
	req := &protov2.RemoveConfigFromAgentGroupRequest{}
	res := &protov2.RemoveConfigFromAgentGroupResponse{}
	req.RequestId = requestId
	req.ConfigName = configName
	req.GroupName = groupName

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}

func GetAppliedAgentGroupForPipelineConfig(requestId []byte, configName string) (*protov2.GetAppliedAgentGroupsResponse, error) {
	url := fmt.Sprintf("%s/GetAppliedAgentGroupsWithPipelineConfig", UserUrlPrefix)
	req := &protov2.GetAppliedAgentGroupsRequest{}
	res := &protov2.GetAppliedAgentGroupsResponse{}
	req.RequestId = requestId
	req.ConfigName = configName

	err := sendProtobufRequest(url, POST, req, res)
	if err != nil {
		return nil, err
	}
	return res, nil
}
